pkg/ingester/ingester_v2.go

```diff
@@ -745,28 +741,24 @@ func (i *Ingester) v2Push(ctx context.Context, req *cortexpb.WriteRequest) (*cor
 
 			// If the cached reference exists, we try to use it.
 			if cachedRefExists {
-				if err = app.AddFast(cachedRef, s.TimestampMs, s.Value); err == nil {
+				var ref uint64
+				if ref, err = app.Append(cachedRef, copiedLabels, s.TimestampMs, s.Value); err == nil {
 					succeededSamplesCount++
+					// This means the reference changes which means we need to update our cache.
+					if ref != cachedRef {
+						db.refCache.SetRef(startAppend, copiedLabels, ref)
+					}
 					continue
 				}
 
-				if errors.Cause(err) == storage.ErrNotFound {
-					cachedRefExists = false
-					err = nil
-				}
-			}
-
-			// If the cached reference doesn't exist, we (re)try without using the reference.
-			if !cachedRefExists {
+			} else {
 				var ref uint64
 
 				// Copy the label set because both TSDB and the cache may retain it.
 				copiedLabels = cortexpb.FromLabelAdaptersToLabelsWithCopy(ts.Labels)
 
-				if ref, err = app.Add(copiedLabels, s.TimestampMs, s.Value); err == nil {
+				if ref, err = app.Append(0, copiedLabels, s.TimestampMs, s.Value); err == nil {
 					db.refCache.SetRef(startAppend, copiedLabels, ref)
-					cachedRef = ref
-					cachedRefExists = true
```

Comment:
Why have you removed this? It was here so that if there are multiple samples for the same series, the next `for _, s := range ts.Samples` iteration will leverage on the ref.

Reply:
done